modules/packages/Image.chpl

```diff
@@ -0,0 +1,322 @@
+  //
+  private inline proc colorOffset(param color) param {
+    return color * bitsPerColor;
+  }
+
+  /* Defines what kind of image to output */
+  enum imageType {
+    /* A BMP image is written from a 2D array of pixels */
+    bmp
+  }
+
+  /*
+    Write an array of pixels as an image
+
+    :arg image: the output filename or an open fileWriter
+    :arg format: the image format
+    :arg pixels: an array of pixel values
+  */
+  proc writeImage(image, format: imageType, pixels: [] pixelType) throws {
+    // the output file channel
+    const outfile = if isSubtype(image.type, fileWriter)
+                      then image
+                      else open(image, ioMode.cw).writer(locking=true);
+    select format {
+      when imageType.bmp do writeImageBMP(outfile, pixels);
+      otherwise
+        halt("Don't know how to write images of type: ", format);
+    }
+  }
+
+  private proc writeImageBMP(outfile, pixels) do
+    compilerError("pixels must be a 2D array of colors");
+
+  private proc writeImageBMP(outfile, pixels: [?d]) throws
+    where d.isRectangular() && d.rank == 2 {
+    const rows = d.dim(0).size,
+          cols = d.dim(1).size,
+          headerSize = 14,
+          dibHeaderSize = 40,  // always use old BITMAPINFOHEADER
+          bitsPerPixel = numColors*bitsPerColor,
+          // row size in bytes. Pad each row out to 4 bytes.
+          rowQuads = divCeil(bitsPerPixel * cols, 32),
+          rowSize = 4 * rowQuads,
+          rowSizeBits = 8 * rowSize,
+          pixelsSize = rowSize * rows,
+          size = headerSize + dibHeaderSize + pixelsSize,
+          offsetToPixelData = headerSize + dibHeaderSize;
+
+    // Write the BMP image header
+    outfile.writef("BM");
+    outfile.writeBinary(size:uint(32), endianness.little);
+    outfile.writeBinary(0:uint(16), endianness.little); /* reserved1 */
+    outfile.writeBinary(0:uint(16), endianness.little); /* reserved2 */
+    outfile.writeBinary(offsetToPixelData:uint(32), endianness.little);
+
+    // Write the DIB header BITMAPINFOHEADER
+    outfile.writeBinary(dibHeaderSize:uint(32), endianness.little);
+    outfile.writeBinary(cols:int(32), endianness.little);
+    outfile.writeBinary(-rows:int(32), endianness.little); /*neg for swap*/
+    outfile.writeBinary(1:uint(16), endianness.little); /* 1 color plane */
+    outfile.writeBinary(bitsPerPixel:uint(16), endianness.little);
+    outfile.writeBinary(0:uint(32), endianness.little); /* no compression */
+    outfile.writeBinary(pixelsSize:uint(32), endianness.little);
+    outfile.writeBinary(2835:uint(32), endianness.little); /*pixels/meter print resolution=72dpi*/
+    outfile.writeBinary(2835:uint(32), endianness.little); /*pixels/meter print resolution=72dpi*/
+    outfile.writeBinary(0:uint(32), endianness.little); /* colors in palette */
+    outfile.writeBinary(0:uint(32), endianness.little); /* "important" colors */
+
+    for i in d.dim(0) {
+      var nbits = 0;
```

Comment:
You do not need to compute this live since it's just bitsPerPixel*cols*numColors